### PR TITLE
Members key share retrieval, decryption and submission 

### PR DESCRIPTION
**Pull request by jumaffre**
Addresses 5. and most of 6. and 6'. in #51 

This PR adds two new member RPC `getEncryptedRecoveryShare` and `submitRecoveryShare` for members to easily retrieve their encrypted shares and submit the then-decrypted share. 

To achieve this, `cchost` now outputs the network public encryption key on startup for members to easily decrypt their shares (using cryptobox). This was slightly painful since `mbedtls` does not yet support x25519 keys and we have to construct the PEM public key manually (see `25519.h`). 

For now, decrypted shares are stored in the node's memory on `submitRecoveryShare` and once the number of submit shares matches the threshold (for now, the number of active members), CCF combines the shares, reconstruct `k_z` and decrypt the ledger secrets in `ccf.shares`. 

There is a new variant of the `recovery_test` e2e test to test the retrieval, decryption (using `pynacl`) and submission of shares. 

Next:
- Plug the share combination into our existing recovery scheme (i.e. members can only submit their shares once the recovery proposal is passed and once the shares have been assembled, the recovery of the private ledger can start).

### CODE REVIEW

File: cmake/common.cmake [hunk 1]
```diff
@@ -175,14 +175,6 @@ else()
   set(TEST_ENCLAVE_TYPE -e virtual)
 endif()
 
-# Test-only option to enable extensive tests
-option(EXTENSIVE_TESTS "Enable extensive tests" OFF)
```
Comment (jumaffre) [inline cmake/common.cmake]: This was unused so I've removed it altogether.

File: src/node/genesisgen.h [hunk 2]
```diff
@@ -269,11 +286,8 @@ namespace ccf
 
     void add_key_share_info(const KeyShareInfo& key_share_info)
     {
-      auto [shares_view, values_view] =
-        tx.get_view(tables.shares, tables.values);
-      auto keyshare_id = get_next_id(values_view, ValueIds::NEXT_KEYSHARE_ID);
-
-      shares_view->put(keyshare_id, key_share_info);
+      auto shares_view = tx.get_view(tables.shares);
```
Comment (jumaffre) [inline src/node/genesisgen.h]: There is no need for a separate key here. We only keep the latest version of the key share in the KV (i.e. key is `0`).

**Comment by cimetrics** [comment]
retrieve_encrypted_share@[5818 aka 20200309.11](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5818&view=results) vs master ewma over 30 builds from [5437](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5437&view=results) to [5804](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5804&view=results)
images

File: src/node/networkencryption.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    NetworkEncryptionKey(bool random = false)
+    {
+      if (random)
+      {
+        private_raw = tls::create_entropy()->random(crypto::BoxKey::KEY_SIZE);
+      }
+    }
+
+    std::vector<uint8_t> get_private_raw()
```
Comment (achamayou) [inline src/node/networkencryption.h]: Why have a getter if the field is public?

Comment (achamayou) [reply src/node/networkencryption.h]: I'm guessing private_raw (:17) is meant to be private.

File: src/node/rpc/memberfrontend.h [hunk 4]
```diff
@@ -411,6 +419,8 @@ namespace ccf
     NetworkTables& network;
     AbstractNodeState& node;
     const lua::TxScriptRunner tsr;
+    // For now, shares are stored in enclave memory
```
Comment (achamayou) [inline src/node/rpc/memberfrontend.h]: I don't think we're going to worry about swapping them out in the foreseeable future.

File: src/node/rpc/memberfrontend.h [hunk 5]
```diff
@@ -684,6 +694,99 @@ namespace ccf
...
+          {
+            return make_error(
+              HTTP_STATUS_BAD_REQUEST, "No key share info available");
+          }
+          std::vector<uint8_t> decrypted_ls(LedgerSecret::MASTER_KEY_SIZE);
+          crypto::GcmCipher encrypted_ls;
+          encrypted_ls.deserialise(key_share_info->encrypted_ledger_secret);
+
+          crypto::KeyAesGcm(share_wrapping_key.data)
+            .decrypt(
```
Comment (achamayou) [inline src/node/rpc/memberfrontend.h]: deserialise and decrypt don't seem to have error handling.

File: src/tls/25519.h [hunk 6]
```diff
@@ -7,71 +7,167 @@
...
+
+      rc = mbedtls_asn1_get_bitstring_null(&p, end, &len);
+      if (rc != 0)
+      {
+        throw std::logic_error(
+          fmt::format("mbedtls_asn1_get_tag failed: {}", error_string(rc)));
+      }
+
+      std::vector<uint8_t> public_raw(p, end);
+      mbedtls_pem_free(&pem);
```
Comment (achamayou) [inline src/tls/25519.h]: leaks if any of the checks throw.

File: src/node/rpc/memberfrontend.h (hunk 4 above)
Comment (jumaffre) [reply src/node/rpc/memberfrontend.h]: Bad phrasing from me: I meant that the submitted shares will be stored in the KV soon. I'll rephrase.

File: src/tls/25519.h (hunk 6 above)
Comment (jumaffre) [reply src/tls/25519.h]: Good spot!

File: src/node/rpc/memberfrontend.h [hunk 7]
```diff
@@ -765,19 +765,24 @@ namespace ccf
           if (!key_share_info.has_value())
           {
             return make_error(
-              HTTP_STATUS_BAD_REQUEST, "No key share info available");
+              HTTP_STATUS_INTERNAL_SERVER_ERROR, "No key share info available");
           }
           std::vector<uint8_t> decrypted_ls(LedgerSecret::MASTER_KEY_SIZE);
           crypto::GcmCipher encrypted_ls;
           encrypted_ls.deserialise(key_share_info->encrypted_ledger_secret);
```
Comment (achamayou) [inline src/node/rpc/memberfrontend.h]: Can the deserialisation also fail?

Comment (jumaffre) [reply src/node/rpc/memberfrontend.h]: I think so but it will be covered by the generic PC exception catch?

Comment (jumaffre) [reply src/node/rpc/memberfrontend.h]: BTW, this code will be moved to a new global hook when the shares get stored in the KV.

**Comment by jumaffre** [comment]
I'll merge this PR now as all the comments have been addressed. The key shares feature is very much work in progress and I expect that most of the code written so far will be reworked, or at least refactored.

**Pull request merged**